### PR TITLE
feat(List): Adds List presentation components GM-185

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

Adds list primitives elements for arbitrary construction.

### Example Usage
```tsx
import { List, ListRow, ListCol } from "@codecademy/gamut";

const View = () => (
  <List variant="slat">
    <ListRow>
      <ListCol type="header" size="md">Row Header</ListCol>
      <ListCol size="lg">Some Row Detail</ListCol> // Implicitly content
     <ListCol type="action"><IconButton /></ListCol>
    </ListRow>
  </List>
);
```

### 

### Principles
- In order to ensure consistent column widths between rows sizes are predefined.
- Primarily functions to list polymorphic entities like users / curriculum / exercises that may have information / interactions that are associated with them.  While statistics / metrics can be displayed in this component and it does its best to behave as a table should it is not the primary function of it.
- Purely presentational.   While we would like to provide a component to manage controlling the data inside the List we do not have enough information to create an API that will serve those needs.

### Components
- `List`:  A list wrapper that creates a new context to share downstream styles with nested components (allowing prop level overrides on nested components).  Comes with 3 stylistic variants: `table`, `slat` and `bar`. 
- `ListRow`: A row component that handles the layout / responsiveness of each constituent column.  At the `xs` breakpoint this will switch from `flex` to `grid` and provide a restructured layout that is agnostic to column configuration.
- `ListCol`: A flex column that comes with predefined column sizes and boolean states for responsiveness.   In addition it includes a variant for 3 possible types `header` `content` and `control` which correspond to the mobile layout.

#### Column Configuration
The 3 type variants have specific effects depending on the top level variant and all effect layout:
- `header`:  Is placed on the top left of the mobile view no matter where it is specified in the markup.  When the `List` is scrollable the element will remain stuck to the left hand border of the list.
- `control`: Is placed at the top right (1/3rd of the row) of the mobile layout.
- `content`: All become full width and are rendered in order below the top row in the mobile layout.

### Behaviors
- Responsiveness:  by default rows will render their columns vertically before 480px.  The top row is split with 2:1 ratio with special column types `header` and `control`.  All other columns render beneath this as 100%.
- `scrollable`:  You can optionally configure lists to keep columns at their defined widths and scroll LTR if there is not enough room, rather than collapsing.  When this is specified the `header` column will lock to the left side and all other content will scroll underneath it.

### Rollout Plan
1. Release to LX to be used in code challenges for their 10/18 layout.  This will act as a base test for the API / Options.
2. Test with teams that have table / list based visuals (Business / Enterprise) to see how well this suits their immediate needs on the presentational level.  At the same time discuss with them access patterns / needs for how they manipulate / control their data.
3. Adjust API / styles to final version.
4. Create proposal for a larger component that provides more logical structure based on the findings of product team user testings.

#### Normal Spacing
Screen Shot 2021-09-08 at 12 52 25 PM

#### Default View
Screen Shot 2021-09-07 at 12 02 27 AM

#### Collapse / Truncated View
Screen Shot 2021-09-07 at 12 02 38 AM

#### Mobile Layout
Screen Shot 2021-09-07 at 12 02 49 AM

## Scrollable List
https://user-images.githubusercontent.com/52927224/132532731-88208f87-50b7-4fb1-b11d-08a65e26a6fd.mov

### PR Checklist

- [x] Related to designs:  [Figma](https://www.figma.com/file/MkQYGsGhMCCUls9TK0pn4U/Lists?node-id=0%3A1)
- [x] Related to JIRA ticket: https://codecademy.atlassian.net/browse/GM-185
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://612f82b101d1f200d8a5a157--gamut-preview.netlify.app>https://612f82b101d1f200d8a5a157--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/612f82b101d1f200d8a5a157>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.40ce64.0<br/>@codecademy/gamut-labs@19.23.14-alpha.40ce64.0<br/>@codecademy/gamut@36.8.2-alpha.40ce64.0<br/>@codecademy/monaco@0.3.5-alpha.40ce64.0<br/>@codecademy/styleguide@43.10.7-alpha.40ce64.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://612f86480a1ca619ab79f12f--gamut-preview.netlify.app>https://612f86480a1ca619ab79f12f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/612f86480a1ca619ab79f12f>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.bb76ab.0<br/>@codecademy/gamut-labs@19.23.14-alpha.bb76ab.0<br/>@codecademy/gamut@36.8.2-alpha.bb76ab.0<br/>@codecademy/monaco@0.3.5-alpha.bb76ab.0<br/>@codecademy/styleguide@43.10.7-alpha.bb76ab.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6130e0b20a73a5ed8d2611bd--gamut-preview.netlify.app>https://6130e0b20a73a5ed8d2611bd--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6130e0b20a73a5ed8d2611bd>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.d93dd1.0<br/>@codecademy/gamut-labs@19.23.14-alpha.d93dd1.0<br/>@codecademy/gamut@36.8.2-alpha.d93dd1.0<br/>@codecademy/monaco@0.3.5-alpha.d93dd1.0<br/>@codecademy/styleguide@43.10.7-alpha.d93dd1.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://613515d3fece25edfbf783b0--gamut-preview.netlify.app>https://613515d3fece25edfbf783b0--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/613515d3fece25edfbf783b0>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.301a1d.0<br/>@codecademy/gamut-labs@19.23.14-alpha.301a1d.0<br/>@codecademy/gamut@36.8.2-alpha.301a1d.0<br/>@codecademy/monaco@0.3.5-alpha.301a1d.0<br/>@codecademy/styleguide@43.10.7-alpha.301a1d.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6136c4290807dde59fc64bd9--gamut-preview.netlify.app>https://6136c4290807dde59fc64bd9--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6136c4290807dde59fc64bd9>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.233124.0<br/>@codecademy/gamut-labs@19.23.14-alpha.233124.0<br/>@codecademy/gamut@36.8.2-alpha.233124.0<br/>@codecademy/monaco@0.3.5-alpha.233124.0<br/>@codecademy/styleguide@43.10.7-alpha.233124.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://61383b012641d62f95dd6ff4--gamut-preview.netlify.app>https://61383b012641d62f95dd6ff4--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/61383b012641d62f95dd6ff4>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.c58b59.0<br/>@codecademy/gamut-labs@19.23.14-alpha.c58b59.0<br/>@codecademy/gamut@36.8.2-alpha.c58b59.0<br/>@codecademy/monaco@0.3.5-alpha.c58b59.0<br/>@codecademy/styleguide@43.10.7-alpha.c58b59.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://61383c9873233d31aa7644eb--gamut-preview.netlify.app>https://61383c9873233d31aa7644eb--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/61383c9873233d31aa7644eb>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.ccdf6e.0<br/>@codecademy/gamut-labs@19.23.14-alpha.ccdf6e.0<br/>@codecademy/gamut@36.8.2-alpha.ccdf6e.0<br/>@codecademy/monaco@0.3.5-alpha.ccdf6e.0<br/>@codecademy/styleguide@43.10.7-alpha.ccdf6e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://613848f17b71e048e4b31d86--gamut-preview.netlify.app>https://613848f17b71e048e4b31d86--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/613848f17b71e048e4b31d86>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.26b135.0<br/>@codecademy/gamut-labs@19.23.14-alpha.26b135.0<br/>@codecademy/gamut@36.8.2-alpha.26b135.0<br/>@codecademy/monaco@0.3.5-alpha.26b135.0<br/>@codecademy/styleguide@43.10.7-alpha.26b135.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6138bf72aa2d2ef140402f44--gamut-preview.netlify.app>https://6138bf72aa2d2ef140402f44--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6138bf72aa2d2ef140402f44>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.51-alpha.2c6ee3.0<br/>@codecademy/gamut-labs@19.23.14-alpha.2c6ee3.0<br/>@codecademy/gamut@36.8.2-alpha.2c6ee3.0<br/>@codecademy/monaco@0.3.5-alpha.2c6ee3.0<br/>@codecademy/styleguide@43.10.7-alpha.2c6ee3.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6138d54f803d79ee419a1af1--gamut-preview.netlify.app>https://6138d54f803d79ee419a1af1--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6138d54f803d79ee419a1af1>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.65-alpha.a3437e.0<br/>@codecademy/gamut-labs@19.24.9-alpha.a3437e.0<br/>@codecademy/gamut@37.1.2-alpha.a3437e.0<br/>@codecademy/monaco@0.3.14-alpha.a3437e.0<br/>@codecademy/styleguide@44.1.5-alpha.a3437e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6138dcdb98d727010400782a--gamut-preview.netlify.app>https://6138dcdb98d727010400782a--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6138dcdb98d727010400782a>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
i'm seeing some weirdness accessibility-wise with between 175% - 200%, i wonder if we should enable scroll on those screen sizes or bump the mobile-view up another view 🤔

File: packages/gamut/src/List/elements.tsx [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    overflowX: 'auto',
+    maxWidth: 1,
+  },
+});
+
+export interface ListProps
+  extends StyleProps<typeof listStates>,
+    StyleProps<typeof listVariants> {}
+
+export const ListEl = styled.div<ListProps>(listVariants, listStates);
```
[the remark this answers is not in the document]
Comment (codecaaron) [reply packages/gamut/src/List/elements.tsx:37]: So I think I'm gonna mess around with the roles a bit.

If we want to render as a table we'll need the TH stuff, but if I swap this to use `ul` and `li` I think we can be in good shape.

[the remark this answers is not in the document]
Comment (dreamwasp) [reply packages/gamut/src/List/elements.tsx:37]: oh duh, RIGHT LISTS, all this table talk has me turned around

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.65-alpha.6a05be.0<br/>@codecademy/gamut-labs@19.24.9-alpha.6a05be.0<br/>@codecademy/gamut@37.1.2-alpha.6a05be.0<br/>@codecademy/monaco@0.3.14-alpha.6a05be.0<br/>@codecademy/styleguide@44.1.5-alpha.6a05be.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://61398af37a3d7073f15d93fd--gamut-preview.netlify.app>https://61398af37a3d7073f15d93fd--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/61398af37a3d7073f15d93fd>Deploy Logs</a></p>

**Comment by songstephen** [comment]
Let's goooo!!! This is so exciting 🤩

This is a classic example of Atomic Design, and I think we should work that into our documentation. We should emphasize that this `List` Organism consists of repeated `ListRow` Molecules that are constructed with `ListCol` Atoms, which can be swapped out or modified for each specific design beyond the templates and screenshots we have.

I think it would help our partner teams think through how to use Lists in their own unique designs, and encourage folks to change the contents and Atoms inside the Lists accordingly. Especially excited for the rollout plan and seeing this come to life!

File: packages/gamut/src/List/__tests__/List-test.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { setupEnzyme } from '@codecademy/gamut-tests';
```
Comment (jakemhiller) [inline packages/gamut/src/List/__tests__/List-test.tsx:1]: y enzyme?

**Review by jakemhiller — APPROVED** [review]

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.75-alpha.5efc48.0<br/>@codecademy/gamut-labs@19.26.4-alpha.5efc48.0<br/>@codecademy/gamut@37.3.3-alpha.5efc48.0<br/>@codecademy/monaco@0.4.3-alpha.5efc48.0<br/>@codecademy/styleguide@44.4.1-alpha.5efc48.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6140b73db02ed258ef3d8dbe--gamut-preview.netlify.app>https://6140b73db02ed258ef3d8dbe--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6140b73db02ed258ef3d8dbe>Deploy Logs</a></p>

**Pull request merged**